tmpl/blog/introducing-charrua-dhcp.md

```diff
@@ -0,0 +1,221 @@
+      | Reply reply ->
+        log c (blue "Received packet %s" (Dhcp_wire.pkt_to_string pkt));
+        N.write net (Dhcp_wire.buf_of_pkt reply)
+        >>= fun () ->
+        log c (blue "Sent reply packet %s" (Dhcp_wire.pkt_to_string reply));
+        Lwt.return_unit
+```
+
+As stated,
+[Dhcp_server.Input.input_pkt](http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt)
+does not perform any IO of its own, it deals only with the logic of inputing a
+DHCP packet and returning a possible answer, which should then be sent by the
+caller. This allows a design where all the side effects are controlled in one
+small chunk, it makes it easier to understand the state transitions when they
+are made explicit.
+
+At the time of this writing,
+[Dhcp_server.Input.input_pkt](http://haesbaert.github.io/charrua-core/api/Dhcp_server.Input.html#VALinput_pkt)
+is not side effect free, as it manipulates a database of leases, this will be
+changed in the next version to be pure as well.
+
+Storing leases in permanent storage is also not supported at this time and
+should be available soon, with irmin and other backends. The main idea is to
+always return a new lease database at every input, or maybe just the updates to
+be applied, in this scenario, the caller would be able to store the database in
+permanent storage as he sees fit.
+
+#### Configuration
+
+This project started independently of MirageOS, and at that time, the best
+configuration I could think of was the well known `ISC` `dhcpd.conf`, therefore
+the configuration uses the same format, but it does not support all the myriad
+of options the original one does.
+
+```ocaml
+  type t = {
+    addresses : (Ipaddr.V4.t * Macaddr.t) list;
+    subnets : subnet list;
+    options : Dhcp_wire.dhcp_option list;
+    hostname : string;
+    default_lease_time : int32;
+    max_lease_time : int32;
+  }
+
+  val parse : string -> (Ipaddr.V4.Prefix.addr * Macaddr.t) list -> t
+  (** [parse cf l] Creates a server configuration by parsing [cf] as an ISC
+      dhcpd.conf file, currently only the options at [sample/dhcpd.conf] are
+      supported. [l] is a list of network addresses, each pair is the output
+      address to be used for building replies and each must match a [network
+      section] of [cf]. A normal usage would be a list of all interfaces
+      configured in the system *)
+```
+
+Although it is a great format, it doesn't play exactly nice with MirageOS and
+OCaml, since the unikernel needs to parse a string at runtime to build the
+configuration, this requires a file IO backend and other complications. The
+next version should provide OCaml helpers for building the configuration, which
+would drop the requirements of a file IO backend and facilitate writing tests.
+
+### Building a simple server
+
+The easiest way is to follow [mirage-skeleton DHCP
+README](https://github.com/mirage/mirage-skeleton/blob/master/dhcp/README.md).
+
```

Comment:
I was expecting this post to give some comparison with existing DHCP implementations — for example in terms of speed, using some [stress](https://github.com/dyninc/DHCPTest-python) [test](http://vigneshmurugesan.com/2013/04/14/running-dhcperf-on-ubuntu/), or in terms of functionality (are there tweaks needed for some clients?).  I do not speak authoritatively but IMHO that would be a nice addition.